### ISSUE TITLE
Device is unregistered but still getting notification

### ISSUE THREAD

**Issue by GokhanArik** [Azure/azure-notificationhubs#31]
Hi,

There is a common problem I saw around. Sometimes a device is registered multiple times. When you unregister it doesn't unregister all registrationIDs associated with the device. It just deletes one. 

The scenario to recreate the issue is that if the user uninstalls our app, we won't be able to unregister their device. When they install again they will be registered twice. If we unregister latest registrationId, the first one still be in the hub.

**Comment by nikitag-ms** [comment]
Is there a chance that somehow the same device token gets re-registered over and over again? Then, indeed, if you just remove one registrations, the others will still stay there. 

There's [a REST API](https://msdn.microsoft.com/en-us/library/dn223271.aspx) (corresponds to getRegistrationsByChannel SDK method) to get all registrations by token. You could use that to see whether there're registrations associated with the device left.

**Comment by GokhanArik** [comment]
I can look at registrations using Visual Studio and delete duplicate ones also. However, for an app that is in production that won't solve the problem. I don't want to write a script to check and remove duplicate entries. Azure SDK should handle that. What do you think?

**Comment by nikitag-ms** [comment]
>  I don't want to write a script to check and remove duplicate entries. Azure SDK should handle that.

Totally agree with that. That's why I was trying to understand why are there duplicates and suggested that maybe the same device token keeps re-registered and gets different registration ids. Are you sure that you only register a single device only once?

**Comment by GokhanArik** [comment]
No, as I said a device gets registered multiple times. Here is a possible scenario,

1. A user installs the app and gets registered to Azure.
2. The user deletes the app. (His registration id is not removed from Notification Hub)
3. The user installs the app again and gets registered to Azure.

Now the user will have two registration IDs for the same device. In production, it might be an edge case. However, when we are testing we keep installing/uninstalling the app and it slows down our testing process.

I think part of the problem is that Google Cloud Messaging is always assigning the same id for the device, so it gets registered multiple times. In my opinion, Azure shouldn't allow duplicate entries for the same PNS handle or make it optional in Notification Hub Settings.

**Comment by GokhanArik** [comment]
@nikitag-ms Is there any solution to that?

**Comment by nikitag-ms** [comment]
Can you store the original registrationId in the local storage that can persist re-installation and then call CreateOrUpdateRegistration?

**Comment by GokhanArik** [comment]
That's what I'm planning to do

**Comment by albfan** [comment]
I am in the same situation. 

Works well
 1. app login register a device with some tags
 2. app logout unregister device correctly

Works wrong
 1. app register a device with some tags
 2. app is closed by user and it is not unregistered
 3. app register a device with other tags (fail 401)
 4. app tryes to unregister all tags with same pnshandle (401 fails)

After receive a 401 (unauthorized) I'm planning to do a unregisterall call, but first task that method do is to refresh pnshandle that gives another 401 so no pnshandle is removed.

I guess the problem is in refreshing registrations.
https://github.com/Azure/azure-notificationhubs/blob/master/Android/notification-hubs-sdk/src/main/java/com/microsoft/windowsazure/messaging/NotificationHub.java#L259

What means exactly 401? I was planning to use directly REST API, but if this can be solved would be great

**Comment by pilondev** [comment]
someone can solve the problem ? i have the same issue 👎

**Comment by VinSmile** [comment]
When adding new registration, you can search for all existing ones that uses the same GCM handle
`
           var oldRegistrations = await hub.GetRegistrationsByChannelAsync(registration.Handle, 10);
`
